### PR TITLE
Add BamNative datatype

### PR DESCRIPTION
**Pull request by bgruening**
* add a new BamNative datatype
  * BamNative will not sort or index the Bam file
* Bam now subclasses from BamNative
* new test tool

### CODE REVIEW

**Comment by mvdbeek** [comment]
Can we make this queryname sorted instead/additionally ? If we can't assume reads are queryname-sorted (i.e as produced by aligners) many tools (duplicate marking, mapq correctors, split read extactors come to mind) would still need to either implement a check to see if the output is readname sorted or worse, sort them regardless of datatype.

**Comment by bgruening** [comment]
Mh, this datatype was meant to not touch the BAM file, take it as it is produced and do not sort at all (and not index), hence the name.

If tools define `bam` as input the converter comes into play and convert `bam_native` files to `bam`, right?

**Comment by mvdbeek** [comment]
I was assuming the point of this datatype is to avoid unnecessary conversions ?!
I.e `paired-end fastq` -> `bwa-mem` (which does coordinate sorting at the end) -> `samblaster` (which does queryname sorting for the input and coordinate sorting for the output). Instead we could directly go `paired-end fastq` -> `bwa-mem` -> `samblaster` and then, if necessary, coordinate sorting.

**Comment by bgruening** [comment]
@mvdbeek it should address the issue in: https://github.com/galaxyproject/tools-iuc/pull/1590/files

`--reorder` will take the order as it is in the input files, so no guarantee that it is name sorted afaik.
But given this PR we probably can add a third type or add a metadata element that indicates the sorting.

**Comment by mvdbeek** [comment]
> --reorder will take the order as it is in the input files, so no guarantee that it is name sorted afaik.

Right, which is the default for most aligners, I believe, but you're right in that this isn't strictly 'SO:queryname'. Maybe some additional metadata is the way to go that indicates that at least pairs of reads are adjacent ? Because that would be good enough for some important tools, but with the BamNative datatype as it is now this could be totally random. So basically only tools that absolutely do not care at all about sort order can consume this datatype, but I'm afraid people could start treating it as a queryname sorted file (because it could be queryname sorted).

**Comment by bgruening** [comment]
@mvdbeek I just tried the following.
```
       <actions>
            <action name="sort_order" type="metadata" default="name_sorted_bag" />
       </actions>
```

With some hacks we could use this and skip/adjust the set_meta function in a way that the index is not created/used. We will probably miss some metadata, but this would enable us to use just one BAM file.
Disadvantage, is that we could still not upload name-sorted BAM files afaik.

ping also @blankenberg to gather more comments

**Comment by mvdbeek** [comment]
> With some hacks we could use this and skip/adjust the set_meta function in a way that the index is not created/used. We will probably miss some metadata, but this would enable us to use just one BAM file.

Yeah, that does look like a good idea. If you know that a tool places reads with the same queryname next to each other (which isn't necessarily what SO:queryname implies) you could use the action to set the metadata accordingly. So in a first instance tool wrappers could handle this logic in the command section, and eventually we could have a syntax to request this format from the framework.

Maybe something like 
```
<param argument="--input" type="data" format="bam,sam" metadata_requirement="adjacent_queryname"/>
```
?

**Comment by bgruening** [comment]
@mvdbeek what about the upload problem? Such "unordered" files could not be uploaded afaik.

Your example would then be already possible by using something like this I think:
```
<param argument="--input" type="data" format="bam,sam" >
     <options options_filter_attribute="metadata.sorted_by" >
          <filter type="add_value" value="adjacent_queryname" />
     </options>
     <validator type="expression" message="">value is not None and value.metadata.sorted == "adjacent_queryname"</validator>
</param>
```

**Comment by blankenberg** [comment]
I like this approach. The default action is to still sniff then sort and index bam files when uploaded by a user, but they can manually upload an un/differently-sorted bam.

**Comment by bgruening** [comment]
@blankenberg do you prefer this more explicit way or the alternative of having one BAM format with metadata indicating the sorting. The latter can probably not uploaded.

**Comment by bgruening** [comment]
@blankenberg or to be more concrete, do you like this PR or the way I sketched out the metadata approach?

**Comment by blankenberg** [comment]
I like this PR, and is what I would have done myself to support unsorted bam. I think the best thing we can probably do at this point is to have the 2+ datatypes. So we can upload un/diff-sorted bam for tools that will need them this way and also don't need to worry about the many existing tools that expect 'bam' to be sorted and indexed and any potential (silent) errors that may be caused.

The use of the metadata field values for limiting inputs can be useful and powerful, but I think they can be a bit clunky to use when developing tools especially in cases when things can be represented by the hierarchical structure of datatypes:
```
bam_native --> [ bam_sorted --> ] sortedBamA(e.g. current BAM),sortedBamB,sortedBamC
```
The hierarchy approach can fall apart a bit when the key/important attribute is greater than one in number. But if sort_order is really the primary defining metadata value, it makes sense to use multiple datatypes.

**Comment by bgruening** [comment]
Ok, thats fine. Than this PR is ready for review. @mvdbeek any objections? Should I add a third datatype as well in this PR?

File: lib/galaxy/datatypes/binary.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: Can we leave this at `BAM files` ? xref https://github.com/galaxyproject/galaxy/pull/5037#discussion_r155796306

File: lib/galaxy/datatypes/binary.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: I'd move that into the base class, and use `return super(Bam, self).sniff(filename) and self.dataset_content_needs_grooming(filename) ` (and maybe change `filename` to `file_name` for consistency ?)

File: test/functional/tools/sam_to_bam_native.xml
Comment (mvdbeek) [inline test/functional/tools/sam_to_bam_native.xml]: Wouldn't this test use the `sam_to_bam` converter ?
Maybe change the input format to `bam_native` ?
If we add this to test/functional/tools/samples_tool_conf.xml it should be run during the functional tests.

File: lib/galaxy/datatypes/binary.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: We don't actually need the index at all for `get_chunk`, so we should probably drop lines 259-263.

File: lib/galaxy/datatypes/binary.py
Comment (bgruening) [reply lib/galaxy/datatypes/binary.py]: Done.

File: lib/galaxy/datatypes/binary.py
Comment (bgruening) [reply lib/galaxy/datatypes/binary.py]: Make sense, but it should be `return super(Bam, self).sniff(filename) and __not__ self.dataset_content_needs_grooming(filename)` right?

File: test/functional/tools/sam_to_bam_native.xml
Comment (bgruening) [reply test/functional/tools/sam_to_bam_native.xml]: I fixed this test, it should test the conversion from SAM to BAM (unsorted).
Thanks @mvdbeek for the review.

File: test/functional/tools/sam_to_bam_native.xml
Comment (mvdbeek) [inline test/functional/tools/sam_to_bam_native.xml]: This still doesn't test the converter, it just runs this tool. You should be able to specify as input format `bam_native` but provide a sam file in the test. Copy that to an output, and the output should be an unsorted bam file.

Comment (bgruening) [reply test/functional/tools/sam_to_bam_native.xml]: Ah, I see. No this test is testing something different. It tests that Galaxy is not sorting this file.
Do you want to see an additional test for the implicit converter as well?

**Comment by mvdbeek** [comment]
@bgruening I added the test I was thinking about in https://github.com/bgruening/galaxy/pull/37.
This points out another issue, which is that the conversion to `bam_native` is going to be `bam` because `Bam` inherits from `BamNative`. I'm just not sure if this is really a problem. After all unsorted may be sorted.

**Comment by bgruening** [comment]
> the conversion to bam_native is going to be bam because Bam inherits from BamNative. 

What do you mean here? The conversion from Bam to BamNative?

> After all unsorted may be sorted.

That is true. A BamNative could be sorted as well. But I would not use the words sorted/unsorted here. Native should just refer to a BAM file that is not touched by Galaxy. Thats by the conversion from Bam to BamNative does not make much sense and has probably no use-case.

**Comment by mvdbeek** [comment]
> Thats by the conversion from Bam to BamNative does not make much sense and has probably no use-case.

Right, but if you use a `sam` dataset to a `bam_native` input, the `sam_to_bam` converter will be used, instead of `sam_to_bam_native`. The `sam_to_bam_native` will actually not be used for implicit conversions right now. But maybe that's fine.

**Comment by bgruening** [comment]
Interesting this seems to be a (unrelated to this PR) bug. If you use this instead:

```
    <datatype extension="sam" type="galaxy.datatypes.tabular:Sam" display_in_upload="true">
        <converter file="sam_to_bam_native.xml" target_datatype="bam_native"/>
        <converter file="sam_to_bam.xml" target_datatype="bam"/>
        <converter file="sam_to_bigwig_converter.xml" target_datatype="bigwig"/>
    </datatype>
```

tt works and sam_to_bam_native is used. So Galaxy seems to take the first in the list and not matches target_datatype.

**Comment by mvdbeek** [comment]
I guess it's not too bad of a mechanism to use the first, given that both are valid targets. If you order it that way and the tests pass that would be good enough for me.

**Comment by bgruening** [comment]
> I guess it's not too bad of a mechanism to use the first, given that both are valid targets. If you order it that way and the tests pass that would be good enough for me.

This would mean that all implicit converters would favor the bam_native conversion.

Please have a look at this PR: https://github.com/bgruening/galaxy/pull/38
This implements at least the behavior I would expect. If the tool wants to have BAM you see `(as bam)` if the tool wants to have BamNative you see now `(as bam_native)`.

This is not implemented to be performant atm, ideally this should be moved into the caching a function up. ideas welcome.

**Comment by mvdbeek** [comment]
> This would mean that all implicit converters would favor the bam_native conversion.

I think that would only happen if `bam_native` and `bam` are valid. Which is `x` to `bam_native` at the moment. `bam` is more restricted, so if  a tool demands `bam` it will get `bam`.

But I can see how it makes sense to prefer an exact match, I'd implement that at `get_converters_by_datatype` though, which is already being cached.

**Comment by bgruening** [comment]
Meh, this looks like a real failing test ...

**Comment by mvdbeek** [comment]
Yeah, I got some local commits for this

**Comment by mvdbeek** [comment]
With https://github.com/bgruening/galaxy/pull/39 the tests should pass, but we'll need to rebase after the merge. I think we can and should do this without the additional sorting for now. 

Otherwise we'd probably have to refine the caching of the results, because the first time we ask for conversions we store the order. Something like `functools.lru_cache` instead of caching the result in `self._converters_by_datatype[ext]` would work, but I'm not convinced we need the added complexity.

**Comment by bgruening** [comment]
@mvdbeek I think I found the difference, try this:

```
        <converter file="sam_to_bam.xml" target_datatype="bam"/>
        <converter file="sam_to_bam_native.xml" target_datatype="bam_native"/>
        <converter file="sam_to_bigwig_converter.xml" target_datatype="bigwig"/>
```
With your PR this will not work. Can you confirm?

**Comment by mvdbeek** [comment]
Right, but why would you do this ?
The datatypes in general are sensitive to the order in which they are defined / listed in the datatypes_conf.xml, so at least that is consistent.

**Comment by bgruening** [comment]
> The datatypes in general are sensitive to the order in which they are defined / listed in the datatypes_conf.xml, so at least that is consistent.

afaik only the sniffers are sensible to the order not the datatypes. I think it is odd if such a functionality is depended on the order if it does not need to. But its your call. Happy to merge it.

**Comment by mvdbeek** [comment]
I'd be more comfortable not changing the way the converters are picked right now. I agree that if you ask for an extension (`bam_native` in this case) you should get that extension and not the extension of a subclass. I'd prefer to work on this in a separate PR if that doesn't bother you too much.

**Comment by bgruening** [comment]
Sure, if you prefer to keep it separate. I have merged your PR and resolved the conflict hopefully.

**Comment by mvdbeek** [comment]
Hmm, those failing tests are passing when I test them locally. I wonder if this is because Jenkins runs an older version of samtools ?

**Comment by mvdbeek** [comment]
Arg, I was thinking we could get around the samtools version problem by using `pysam.view`, but the implementation is only partial and doesn't support writing to a file. So what now ? Upgrading samtools in the testing container would be my next suggestion ...

**Comment by nsoranzo** [comment]
Maybe the failing framework tests can be fixed by auto-installing the samtools requirements with `conda_auto_install = True` option in `config/galaxy.ini`?

**Comment by mvdbeek** [comment]
We should be eating our own dogfood, right ? :D

**Comment by martenson** [comment]
I wonder how much slower it will make the tests.

**Comment by mvdbeek** [comment]
Shouldn't be much slower, around a minute or so.

File: test/base/driver_util.py
Comment (nsoranzo) [inline test/base/driver_util.py]: Should these have value `getattr(config_object, 'conda_auto_init', False)` and `getattr(config_object, 'conda_auto_install', False)` respectively?
If yes, then probably the new lines 799-800 are not needed.

Comment (mvdbeek) [reply test/base/driver_util.py]: Yes, that should work!

**Comment by bgruening** [comment]
Cool, thanks @mvdbeek seems to work!

**Review by mvdbeek — APPROVED** [review]
Looks good to me! Keep in mind that BamNative makes no guarantees to sort oder at all. 

We should (in another PR) implement some variants of this for the cases where reads are in the order that has been spit out by the aligner (`BamReadsAdjacent` ?) and where reads are ordered by queryname (`BamQueryNameSorted` ?) and by tag (`BamTagSored` ?).

**Comment by mvdbeek** [comment]
Thanks a lot @bgruening !

**Comment by bgruening** [comment]
Thanks to you @mvdbeek!
Do you want me to get this ordering of datatypes in or not?

**Comment by martenson** [comment]
>We should (in another PR) implement some variants of this for the cases where reads are in the order that has been spit out by the aligner (BamReadsAdjacent ?) and where reads are ordered by queryname (BamQueryNameSorted ?) and by tag (BamTagSored ?).

This warranties an issue I think

**Pull request merged**
